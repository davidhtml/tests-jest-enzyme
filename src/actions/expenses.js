import uuid from 'uuid';
import database from '../firebase/firebase';
//ADD_EXPENSE
const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

//startADD expense
const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;

        const expense = {
            description,
            note,
            amount,
            createdAt
        }

        return database.ref(`users/${uid}/expenses`).push(expense)
                .then((ref) => {
                    dispatch(addExpense({
                        id: ref.key,
                        ...expense
                    }))
                })
    }
}

// REMOVE_EXPENSE
const removeExpense = ({ id } = {}) => {
    return ({
        type: 'REMOVE_EXPENSE',
        id
    })
}

const startRemoveExpense = ({id} = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).remove()
            .then(() => {
                dispatch(removeExpense({ id }));
            });
    };
};

// EDIT_EXPENSE
const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates,
})

const startEditExpense = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).update({
        ...updates
        })
            .then(() => {
                dispatch(editExpense(id, updates))
        })
    }
}

//setting Expenses

const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses,
});

const startSetExpenses = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses`).once('value')
                .then((snapshot) => {
                    const expenses = [];
                    snapshot.forEach((childSnapshot) => {
                        expenses.push({
                            id: childSnapshot.key,
                            ...childSnapshot.val()
                        })
                    })
                    dispatch(setExpenses(expenses))
                })
    }


    ///Fetch all expense data once
    //// Parse that data into array
    //// Dispatch SET_EXPENSES


};

export {addExpense, removeExpense, editExpense, startAddExpense, setExpenses, startSetExpenses, startRemoveExpense, startEditExpense};
