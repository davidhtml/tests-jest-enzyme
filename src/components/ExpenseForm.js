import React from 'react';
import moment from "moment";
import {SingleDatePicker} from 'react-dates';
import 'react-dates/initialize'

class ExpenseForm extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            description: '',
            note: '',
            amount: '',
            createdAt: moment(),
            calendarFocused: null,
            errorMessage: null
        }
    }

    componentDidMount() {
        const {existingExpense} = this.props;
        if (existingExpense) {
            this.setState(() => {
                return {
                    description: existingExpense.description,
                    amount: (existingExpense.amount / 100).toString(),
                    note: existingExpense.note,
                    createdAt: moment(existingExpense.createdAt)
                }
            })
        }
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;

        this.setState(() => ({ description: description}))
    }

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({note: note}))
    }

    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount: amount }))
        }
    }

    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({createdAt: createdAt}))
        }

    }

    onFocusChange = (brazauskas) => {
        this.setState(() => ({ calendarFocused: brazauskas.focused }))
    }

    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            let msg = 'Please provide the description and the amount';
            this.setState(() => ({errorMessage: msg}))
        } else {
            this.setState(() => ({errorMessage: null}))
            this.props.stateObjToParent({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10)*100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note

            });
        }
    }

    render() {
        return (

            <div>
                {this.state.errorMessage && <h4>{this.state.errorMessage}</h4>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type="text"
                        placeholder="amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />

                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        id={this.state.description}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />


                    <textarea
                        placeholder="add a note for your expense"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                        >
                    </textarea>
                    <button>Add expense</button>
                </form>
            </div>

        );
    }
}

export default ExpenseForm;
