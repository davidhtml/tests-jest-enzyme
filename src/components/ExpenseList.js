import React from 'react';
import {connect} from 'react-redux'
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

const ExpenseList = (props) => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-desktop">Expense</div>
            <div className="show-for-desktop">Amount</div>
        </div>
        <div className="list-body">
            {
                props.expenses.length === 0
                ? (
                    <div className="list-item--message">
                        You have no expenses
                    </div>
                )
                : (
                    props.expenses.map((expense) => {
                        return <ExpenseListItem key={expense.id} {...expense} />
                    })
                )
            }
        </div>
    </div>
);


const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}


export { ExpenseList };
export default connect(mapStateToProps)(ExpenseList);;
