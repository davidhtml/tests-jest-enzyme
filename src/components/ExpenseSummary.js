import React from 'react';
import {connect} from 'react-redux';
import selectExpenses from '../selectors/expenses'
import expensesTotal from '../selectors/expenses-total';
import numeral from 'numeral'

class ExpenseSummary extends React.Component {

    renderExpenses = () => {
        const total = expensesTotal(this.props.expenses);
        return (
            <p>total of {this.props.expenses.length} expenses to pay {numeral(total/100).format('$0,0.00')}</p>
        )
    }

    render() {
        return (
            <div>
                <h4>Here will be how many visible expenses we do have and what is total amount</h4>
                {this.props.expenses.length > 0 && this.renderExpenses()}
            </div>
        );
    }
}

const mapStateToProps = state => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}

export {ExpenseSummary}
export default connect(mapStateToProps)(ExpenseSummary);
