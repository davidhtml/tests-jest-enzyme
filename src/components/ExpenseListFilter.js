import React from 'react';
import {connect} from 'react-redux';
import {setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate} from '../actions/filters';
import {DateRangePicker} from 'react-dates';
import 'react-dates/initialize';

class ExpenseListFilters extends React.Component {

    state = {
        calendarFocused: null,
    }

    onDatesChange = (obj) => {
        this.props.setStartDate(obj.startDate);
        this.props.setEndDate(obj.endDate);
    }

    onFocusChange = (calendarFocused) => {
        console.log(calendarFocused)
        this.setState(() => ({ calendarFocused }))
    }

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    }

    onSortChange = (e) => {
        const option = e.target.value;
        if (option === 'date') {
            this.props.sortByDate();
        } else if (option === 'amount') {
            this.props.sortByAmount();
        }
    }

    render() {
        // console.log(this.props.filters.endDate)
        const {dispatch} = this.props;
        return (
            <div>
                <input
                    type="text"
                    value={this.props.filters.text}
                    onChange={this.onTextChange}/>
                <select
                    value={this.props.filters.sortBy}
                    onChange={this.onSortChange}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    startDateId={"start"}
                    endDate={this.props.filters.endDate}
                    endDateId={"end"}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    showClearDates={true}
                 />
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters,
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        setTextFilter: (text) => dispatch(setTextFilter(text)),
        sortByDate: () => dispatch(sortByDate()),
        sortByAmount: () => dispatch(sortByAmount()),
        setStartDate: (startDate) => dispatch(setStartDate(startDate)),
        setEndDate: (endDate) => dispatch(setEndDate(endDate))
    }
}

export { ExpenseListFilters };
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
