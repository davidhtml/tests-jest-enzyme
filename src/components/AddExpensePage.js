import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';


class AddExpensePage extends React.Component {

    onSubmit = (expense) => {
        this.props.addExpense(expense)
        this.props.history.push('/');
    }

    render() {
        return (
            <div>
                <h1>Add expense</h1>
                <ExpenseForm
                    stateObjToParent={this.onSubmit}
                 />
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        addExpense: (expense) => dispatch(addExpense(expense))
    };
};

export {AddExpensePage};

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
