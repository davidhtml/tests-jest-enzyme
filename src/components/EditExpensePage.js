import React from 'react';
import {connect} from 'react-redux'
import { editExpense, removeExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

//refactor to class bsaed compoennt
// create mapDispatchToProps

//should render addExpense page
// should handle editExpense using spies
//should handle removeExpense using spies


class EditExpensePage extends React.Component {

    onEditExpense = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    }

    onRemove = () => {
        this.props.removeExpense({id: this.props.expense.id});
        this.props.history.push('/');
    }

    render() {
        return (
            <div>
                <ExpenseForm
                    stateObjToParent={this.onEditExpense}
                    existingExpense={this.props.expense}
                 />
                 <button
                     onClick={this.onRemove}
                >
                     Remove expense
                 </button>
            </div>
        );
    }

}



const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        editExpense: (id, expense) => dispatch(editExpense(id, expense)),
        removeExpense: (id) => dispatch(removeExpense(id))
    };
};

export {EditExpensePage};
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
