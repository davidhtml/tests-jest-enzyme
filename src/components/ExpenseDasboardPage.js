import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilter';
import ExpenseSummary from './ExpenseSummary';

const ExpenseDasboardPage = () => (
    <div>
        <ExpenseSummary />
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default ExpenseDasboardPage;
