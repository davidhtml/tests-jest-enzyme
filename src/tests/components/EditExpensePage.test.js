import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

    let startEditExpense, startRemoveExpense, history, wrapper;

    beforeEach(() => {
     startEditExpense = jest.fn();
     startRemoveExpense = jest.fn();
     history = { push: jest.fn()};
     wrapper = shallow(<EditExpensePage
         startEditExpense={startEditExpense}
         startRemoveExpense={startRemoveExpense}
         history={history}
        expense={expenses[0]}
     />);
    })

test('should render edit expense page correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle edit expense', () => {
    wrapper.find('ExpenseForm').prop('stateObjToParent')(expenses[0]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startEditExpense).toHaveBeenLastCalledWith(expenses[0].id, expenses[0]);

});


test('should handle remove expense', () => {
    wrapper.find('button').simulate('click');
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startRemoveExpense).toHaveBeenLastCalledWith({id: expenses[0].id});

});

//
// wrapper.find('form').simulate('submit', {
//     preventDefault: () => {}
// });
