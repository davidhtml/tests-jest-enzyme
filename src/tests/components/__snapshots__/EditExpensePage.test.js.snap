// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render edit expense page correctly 1`] = `
<div>
  <ExpenseForm
    existingExpense={
      Object {
        "amount": 195,
        "createdAt": 0,
        "description": "rent",
        "id": "1",
        "note": "first note",
      }
    }
    stateObjToParent={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove expense
  </button>
</div>
`;
